### DISCUSSION TITLE
How best to stopPropagation/preventDefault?

### DISCUSSION THREAD

**Discussion by tslocke** [solidjs/solid#977] [Q&A]
Say I have a component that delegates an event to some prop, e.g.

```jsx
const Comp = (props) => <div onClick={c.action} ... />
```

and I want to always stopPropagation and preventDefault before delegating (or after, I guess).

What's the best way to do that?

If I knew `props.action` was a function I would just

```jsx
onClick=(e => {e.preventDefault(); e.stopPropagation(); props.action(e)})
```

But this doesn't work if `props.action` is `[fn, arg]`

For now I've added an extra wrapper `<div onClick={e => { e.stopPropagation(); e.preventDefault() }}>` but I'd rather avoid that.

Thanks

**Comment by atk** [comment]
You can simple check if props.action is an array:

```ts
onCLick={e => e.preventDefault(); e.stopPropagation(); Array.isArraqy(props.action) ? props.action[1](props.action[0], e) : props.action(e)}
```

**Reply by tslocke** [reply]
I think this undermines the benefits of the array style right? e.g. if you do a lot of these in a loop, you are creating a ton of closures.

> Solid also supports passing an array to the event handler to bind a value to the first argument of the event handler. This doesn't use bind or create an additional closure, so it is a highly optimized way of delegating events.

**Comment by yw662** [comment]
I would say, 
```ts
C: Component<{onClick: () => void}> = props => <... onClick={
    e => {
        e.preventDefault()
        e.stopPropagation()
        props.onClick()
        return false
    }
}.../>
...

<C onClick={()=>action(...actionArgs)} />
```

**Comment by Exelord** [comment]
That seams like something that could be handled by namespaces and event modifiers (not implemented yet)
```
on:click$preventDefault={[fn, arg]}
on:click$preventDefault={(e) => console.log(e)}
```

Wondering what's the Svelte Devs experience by using it. For me it it looks like much more cleaner. 

Alternatively I would opt in into [custom directive](https://www.solidjs.com/docs/latest/api#use%3A___):

```
use:on={["click", action, { prevent: true }]}
```

**Reply by tslocke** [reply]
Nice - I hadn't come across custom directives. There's a few things I'm missing from Vue that could be supported that way.
